### PR TITLE
Replacing python cosmology component by TensorFlow

### PR DESCRIPTION
**Pull request by dlanzieri**
We implemented tfbackground and some tests . This pullrequest solves the issues #46

### CODE REVIEW

**Comment by dlanzieri** [comment]
Test has failed because of the missing TensorFlow Probability @EiffL

**Comment by EiffL** [comment]
Hi @dlanzieri , I've found the reason why your tests were never returning :-) It's because you had some plotting functions in the `tfbackground.py` file. Everytime you import that tfbackground like this:
```python
import flowpm.tfbackground as tfbackground
```
it runs all of the commands it finds in the file, so in your case, it was running the plotting functions. In general you only want to have functions and/or classes definitions in a module file like tfbackground.py.

You can still add a section at the bottom of the file with plotting instructions to make some plots/tests, but you want to add a  declaration like this before:
```python
if __name__ == "__main__":
```
This tells Python to only run these instructions if you are trying to execute the file like a script, i.e. run `python flowpm/tfbackground.py`, and these instructions will be ignored if you simply import the file.

I have made a small modification on my last commit to move the plotting instructions under this section.  I have also added the necessary things to the `setup.py` to make sure tensorflow_probability gets installed automatically for the tests on travis CI.

You can see the contents of commits by clicking on the last commit above named `Fix testing issues`, or by clicking [here](https://github.com/modichirag/flowpm/pull/56/commits/3dfc6bad346e6e923f1c5d2877d46c47382e8624)

All the tests are working on my machine :-D
```
[francois@Mizar flowpm]$ pytest 
========================================= test session starts ==========================================
platform linux -- Python 3.8.6, pytest-6.1.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/francois/repo/flowpm, configfile: pytest.ini
collected 14 items                                                                                     

tests/test_tfpm.py ......                                                                        [ 42%]
tests/test_utils.py ...                                                                          [ 64%]
tests/tf_background_test.py .....                                                                [100%]

========================================= 14 passed in 11.87s ==========================================
```

To make sure that you have all of the necessary tools installed on your computer to run all the tests, you should try to reintstall flowpm with the following command:
```bash
$ pip install --user -e .[testing]
```
note the `[testing]` section, which tells pip to install the dependencies necessary for testing, like the fastpm code we are using to compare to.

File: flowpm/tfbackground.py [hunk 1]
```diff
@@ -0,0 +1,694 @@
...
+    """
+    a=tf.convert_to_tensor(a,dtype=tf.float32)
+    # Maybe compute ODE or use stored result
+    cache = maybe_compute_ODE(cosmo)
+    lna = tf.math.log(a)
+    return tfp.math.interp_regular_1d_grid(lna,
+                      tf.math.log(cache['a'][0]), tf.math.log(cache['a'][-1]), cache['D2f'])
+
+def f1(cosmo, a):
+    a=tf.convert_to_tensor(a,dtype=tf.float32)
```
Comment (EiffL) [inline flowpm/tfbackground.py:497]: missing documentation

File: flowpm/tfbackground.py [hunk 2]
```diff
@@ -0,0 +1,694 @@
...
+    return D1f(cosmo, a) * a / D1(cosmo, a)
+
+def f2(cosmo, a):
+    a=tf.convert_to_tensor(a,dtype=tf.float32)
```
Comment (EiffL) [inline flowpm/tfbackground.py:501]: missing documentation

File: flowpm/tfbackground.py [hunk 3]
```diff
@@ -0,0 +1,694 @@
...
+    return D2f(cosmo, a)  * a / D2(cosmo, a)
+
+###############################################################
+def Gp(cosmo, a):
```
Comment (EiffL) [inline flowpm/tfbackground.py:505]: This function is redundant with D1 :-/ we should remove one of them

File: flowpm/tfbackground.py [hunk 4]
```diff
@@ -0,0 +1,694 @@
...
+
+    .. math::
+        Gp(cosmo,a)=D'_{1norm}*a
+    """
+    a=tf.convert_to_tensor(a,dtype=tf.float32)
+    return D1(cosmo, a)
+################################################################
+def Gf(cosmo, a):
+    """
+    FastPM growth factor function
```
Comment (EiffL) [inline flowpm/tfbackground.py:532]: what is the difference with Gp? Needs better documentattion

File: flowpm/tfbackground.py [hunk 5]
```diff
@@ -0,0 +1,694 @@
...
+        Gf_2(a)=D'_{2norm}*a**3*E(a)
+    """
+    a=tf.convert_to_tensor(a,dtype=tf.float32)
+    return D2f(cosmo,a)*a**3*E(cosmo,a)
+
+################################################################
+
+def gf(cosmo, a):
+    """
+    Derivative of Gf against a
```
Comment (EiffL) [inline flowpm/tfbackground.py:582]: This should have a better description that doesnt reference another function

File: flowpm/tfbackground.py [hunk 6]
```diff
@@ -0,0 +1,694 @@
...
+
+            Parameters
+            ----------
+            a : tf.TensorArray
```
Comment (EiffL) [inline flowpm/tfbackground.py:586]: missing cosmo

File: flowpm/tfbackground.py [hunk 7]
```diff
@@ -0,0 +1,694 @@
...
+    lna = tf.math.log(a)
+    d1f = tfp.math.interp_regular_1d_grid(lna,
+                      tf.math.log(cache['a'][0]), tf.math.log(cache['a'][-1]), cache['D1f'])
+    f1p = tfp.math.interp_regular_1d_grid(lna,
+                      tf.math.log(cache['a'][0]), tf.math.log(cache['a'][-1]), cache['F1p'])
+    return  (f1p * a ** 3 *E(cosmo,a) +  d1f*a ** 3 * dEa(cosmo,a)  +   3 * a ** 2 * E(cosmo,a)*d1f)
+
+def gf2(cosmo, a):
+    """
+    Derivative of Gf2 against a
```
Comment (EiffL) [inline flowpm/tfbackground.py:615]: same comment as above, needs better description

File: flowpm/tfbackground.py [hunk 8]
```diff
@@ -0,0 +1,694 @@
...
+
+            Parameters
+            ----------
+            a : tf.TensorArray
```
Comment (EiffL) [inline flowpm/tfbackground.py:619]: missing cosmo

File: flowpm/tfbackground.py [hunk 9]
```diff
@@ -0,0 +1,694 @@
...
+    cache = maybe_compute_ODE(cosmo)
+    lna = tf.math.log(a)
+    d2f = tfp.math.interp_regular_1d_grid(lna,
+                      tf.math.log(cache['a'][0]), tf.math.log(cache['a'][-1]), cache['D2f'])
+    f2p = tfp.math.interp_regular_1d_grid(lna,
+                      tf.math.log(cache['a'][0]), tf.math.log(cache['a'][-1]), cache['F2p'])
+    return  (f2p * a ** 3 *E(cosmo,a) +  d2f*a ** 3 * dEa(cosmo,a) + 3 * a ** 2 * E(cosmo,a)*d2f)
+
+#
+# if __name__ == "__main__":
```
Comment (EiffL) [inline flowpm/tfbackground.py:647]: We can keep this main section and use it to plot a few things for tests if that's useful. If not useful, we sshould remove all of this commented section

File: flowpm/tfpower.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+    pknorm = cosmo["sigma8"] ** 2 / sigmasqr(cosmo, 8.0, transfer_fn, **kwargs)
+
+    pk = primordial_matter_power(cosmo, k) * t ** 2 * g ** 2
+
+    # Apply normalisation
+    pk = pk * pknorm
+    return tf.squeeze(pk)
+
+def sigmasqr(cosmo, R, transfer_fn, kmin=0.0001, kmax=1000.0, ksteps=5, **kwargs):
```
Comment (EiffL) [inline flowpm/tfpower.py:200]: either hide that function from imports.... or improve documentation

File: setup.py [hunk 11]
```diff
@@ -13,11 +14,11 @@
...
       license='MIT',
-      packages=['flowpm'],
-      install_requires=['astropy', 'scipy', 'mesh-tensorflow'],
-      tests_require=['fastpm'],
+      packages=find_packages(),
+      install_requires=['astropy', 'scipy', 'mesh-tensorflow', 'tensorflow_probability'],
+      tests_require=['fastpm', 'pyccl'],
       extras_require={
-        'testing':  ["fastpm"],
+        'testing':  ['fastpm', 'pyccl'],
```
Comment (EiffL) [inline setup.py:21]: we should try to remove this dependency on pyccl, nbodykit has functions for cosmology computations, we can use those

File: tests/test_tfpower.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
+# Tests analytic power spectra computations
+import tensorflow as tf
+import numpy as np
+import flowpm.tfbackground as bkgrd
+import flowpm.tfpower as power
+import pyccl as ccl
```
Comment (EiffL) [inline tests/test_tfpower.py:6]: replace pyccl by nbodykit

**Review by EiffL — COMMENTED** [review]
I've done a full pass through the PR and left a few comments, it's looking good so far :-) Most of my comments are related to the documentation. 
One thing I notice is that we have a lot of functions like Gp gp D1, etc with names that are not explicit and these functions are even sometimes redundant, we should try to see what is the mininum set of funcntions we need for flowpm

File: setup.py (hunk 11 above)
Comment (EiffL) [reply setup.py]: removed dependency on pyccl and switched to nbodykit

File: tests/test_tfpower.py (hunk 12 above)
Comment (EiffL) [reply tests/test_tfpower.py]: done as well

**Comment by EiffL** [comment]
Thanks for the updates @dlanzieri  :-) When you address a comment, you can leave a message in the "Reply" box below the comment, and if you believe the comment is fully addressed you can "resolve conversation", if you are not sure that you can fully address the comment, or if you disagree, you can respond to it and we only resolve the conversation when we both agree on the comment.

I have just addressed one of my comments which was about not using `pyccl` as a reference for testing the Eisenstein Hu power spectrum, and instead I'm now using `nbodykit` which is already a dependency of the code.

**Comment by EiffL** [comment]
oups, looks like your tests are failing ^^^^

**Comment by EiffL** [comment]
Fantastic!!! I just had a look and it looks great! So happy to see all of the pieces come together!

I have a few additional suggestions:

- [ ] : you can completely replace the old `tfpm.py` with the content of your `denise_tfpm.py` and you can keep the same function names. Don't worry about replacing things :-)
The code in `tfpm.py` is being tested against an external reference (python-fastpm) in this test:
https://github.com/modichirag/flowpm/blob/tfcosmo/tests/test_tfpm.py
So you can modify the content of `tfpm.py`, and these tests will tell you if you have broken something. You can see how these tests are coded, and you can try to run your own fastpm simulation if you want to have a point of reference of what a "correct code" does.

- [ ] : you can rename the old `baground.py` as `legacy_background.py` to make clear that it's not used for anything anymore, except maybe testing.

- [ ] : I see that your commit messages are a bit long, for instance, this is what appears on github:
image
The idea is that commit messages can be kept short, they should be about what this commit changes in the code and you can commit often so that you don't have one massive commit.
You can provide more details about the modifications you have made in this PR thread for instance. 

- [ ] : You can see if you go this PR thread on GitHub that the automated tests are failing after your commit from yesterday. It appears in several places, next to the PR name in the list:
image
(it's the red cross)
You can see it after each individual commits if they pass the tests or not in this thread:
image
For instance you see there that I made a bunch of broken commits before finding how to fix the bug.
And you can see it at the end of the thread there:
image
This is the place you want to look at to find out what's wrong, if you click on:
- `Details`
- Then on the `The build` link here:
image
It will send you to Travis CI, and you will see a page like this of Travis trying to run the tests:
image
and here we see that the tests are failing because you have removed the Gp function.

Sorry for the long message ^^' Just wanted to give you a few extra-tips on GitHub

File: flowpm/tfpm.py [hunk 13]
```diff
@@ -145,7 +138,7 @@ def lpt2_source(dlin_k, kvec=None, name="LPT2Source"):
     source = tf.multiply(source, 3.0/7.)
     return r2c3d(source, norm=nc**3)
 
-def lpt_init(linear, a0, order=2, cosmology=Planck15, kvec=None, name="LPTInit"):
+def lpt_init(linear, a, af,n_steps,order=2, cosmology=cosmo, kvec=None, name="LPTInit"):
```
Comment (EiffL) [inline flowpm/tfpm.py:141]: This is not correct! `lpt_init` does not have af and n_steps parameters

File: flowpm/tfpm.py [hunk 14]
```diff
@@ -164,25 +157,28 @@ def lpt_init(linear, a0, order=2, cosmology=Planck15, kvec=None, name="LPTInit")
     Q = np.repeat(Q.reshape((1, -1, 3)), batch_size, axis=0)
     pos = Q
 
-    a = a0
-
+    a = a
```
Comment (EiffL) [inline flowpm/tfpm.py:160]: we can remove this line, it's not doing anything

File: flowpm/tfpm.py [hunk 15]
```diff
@@ -164,25 +157,28 @@ def lpt_init(linear, a0, order=2, cosmology=Planck15, kvec=None, name="LPTInit")
...
+    
     lineark = r2c3d(linear, norm=nc**3)
 
-    pt = PerturbationGrowth(cosmology, a=[a], a_normalize=1.0)
-    DX = tf.multiply(dtype(pt.D1(a)) , lpt1(lineark, pos))
-    P = tf.multiply(dtype(a ** 2 * pt.f1(a) * pt.E(a)) , DX)
-    F = tf.multiply(dtype(a ** 2 * pt.E(a) * pt.gf(a) / pt.D1(a)) , DX)
+    #pt = PerturbationGrowth(cosmology, a=[a], a_normalize=1.0)
```
Comment (EiffL) [inline flowpm/tfpm.py:164]: remove this line

File: flowpm/tfpm.py [hunk 16]
```diff
@@ -164,25 +157,28 @@ def lpt_init(linear, a0, order=2, cosmology=Planck15, kvec=None, name="LPTInit")
...
+      P2 = tf.multiply(a ** 2 * f2(cosmo,a) * E(cosmo,a) , DX2)
+      F2 = tf.multiply(a ** 2 * E(cosmo,a) * gf2(cosmo,a) /D2(cosmo,a) , DX2)
       DX = tf.add(DX, DX2)
       P = tf.add(P, P2)
       F = tf.add(F, F2)
 
     X = tf.add(DX, Q)
     return tf.stack((X, P, F), axis=0)
 
+    X = tf.add(DX, Q)
```
Comment (EiffL) [inline flowpm/tfpm.py:179]: what is this doing here ^^' ? Looks like a copy/paste of the same lines above

File: flowpm/tfpm.py [hunk 17]
```diff
@@ -253,16 +249,15 @@ def drift(state, ai, ac, af, cosmology=Planck15, dtype=np.float32,
   with tf.name_scope(name):
     state = tf.convert_to_tensor(state, name="state")
 
-    pt = PerturbationGrowth(cosmology, a=[ai, ac, af], a_normalize=1.0)
-    fac = 1. / (ac ** 3 * pt.E(ac)) * (pt.Gp(af) - pt.Gp(ai)) / pt.gp(ac)
+    fac = 1. / (ac ** 3 * E(cosmo,ac)) * (D1(cosmo,af) - D1(cosmo,ai)) / D1f(cosmo,ac)
```
Comment (EiffL) [inline flowpm/tfpm.py:252]: ok great, but does that mean we don't need the Gp and gp functions in `background.py`?

File: tests/test_tfpm.py [hunk 18]
```diff
@@ -45,7 +47,7 @@ def test_lpt_init():
 
   # Same thing with flowpm
   tlinear = tf.expand_dims(np.array(lineark.c2r()), 0)
-  tfread = tfpm.lpt_init(tlinear, a0, order=1).numpy()
+  tfread = tfpm.lpt_init(tlinear, a0, af,n_steps, order=1).numpy()
```
Comment (EiffL) [inline tests/test_tfpm.py:50]: so here we should remove af and n_steps

File: tests/test_tfpm.py [hunk 19]
```diff
@@ -33,6 +33,8 @@ def test_lpt_init():
   Checking lpt init
   """
   a0 = 0.1
+  af = 1.0
```
Comment (EiffL) [inline tests/test_tfpm.py:36]: also don't need this here

File: tests/test_tfpm.py [hunk 20]
```diff
@@ -107,6 +109,8 @@ def test_nody():
   """ Checking end to end nbody
   """
   a0 = 0.1
+  af = 1.0
+  n_steps=10
```
Comment (EiffL) [inline tests/test_tfpm.py:113]: same. let's remove this

File: tests/test_tfpm.py [hunk 21]
```diff
@@ -122,7 +126,7 @@ def test_nody():
 
   # Same thing with flowpm
   tlinear = tf.expand_dims(np.array(lineark.c2r()), 0)
-  state = tfpm.lpt_init(tlinear, a0, order=1)
+  state = tfpm.lpt_init(tlinear, a0, af,n_steps, order=1)
```
Comment (EiffL) [inline tests/test_tfpm.py:129]: remove af and nsteps

**Review by EiffL — CHANGES_REQUESTED** [review]
I have noticed a few mistakes, nothing major, but otherwise looks good :-)

**Review by EiffL — COMMENTED** [review]
Looking good, ready for review I think

**Comment by EiffL** [comment]
Hi @modichirag ! could you take a look at this PR and tell us what you think? It should replace all of the old python code with a cool new Tensorflow implementation of all cosmology functions!

**Review by EiffL — APPROVED** [review]

**Comment by modichirag** [comment]
This is brilliant. Thank you @dlanzieri and @EiffL for doing this. I have 2 comments that I would like your opinion on before we merge this (though it can be merged now as well) -

1. the major difference from previous version in terms of UI I see is that some tfpm functions take in a cosmology dictionary.
Of all tfpm functions, I think - nbody, lpt_init and linear_field will be called externally while others will likely not be (atleast for an end-to-end pm simulation). Of these, only linear_field does not take in the cosmo dict, but a interpolated 'pk'. My question is this - do we want to make 'pk' part of the cosmology object so that all three functions have consistent signature in terms of comsology dependence? It might also be easier in terms of implementing differentiability w.r.t. cosmology parameters since pk will also depend on them, though then we might be constrained to use Eisenstein Hu? 
So maybe we can keep things as they are right now, giving more flexibility of providing a different interpolated power spectrum instead of constraining ourselves.
What do you guys think?

2. do we want to check for differentiability with respected to some cosmology parameters in one of the tests before merging or should we work on that after things are setup, and that the gradients match? No reason why they would not though, so maybe this is not required at this stage. 

Thanks again for this extensive PR.

**Review by modichirag — COMMENTED** [review]
Looks good :)

**Pull request merged**